remove draftd

bacongobbler opened the pull request:
This PR removes Draftd from the cluster and runs builds locally. This is probably one of the most significant changes ever made to Draft since the switch from websockets to gRPC, so hold onto your nickers!

![ocean's eleven - the pinch](https://i.makeagif.com/media/4-02-2015/icjHZi.gif)

The most noteworthy information:

- draftd has been removed
- `draft init` serverside bootstrapping for draftd has been removed
- ingress configuration has been removed, since we have `draft connect` now and there's no need to set a `basedomain`
- ~~`draft create` now has a `--repository` flag, which is the name of the Docker registry to publish the image to (see the getting started guide in this PR for more info)~~ This has been replaced with `draft config set registry` and the `registry` flag in draft.toml
- Draft plugin's `useTunnel` now initiates a connection to Tiller instead of Draft
- re-named all `$DRAFT_NAMESPACE` and `--draft-namespace` flags to `$TILLER_NAMESPACE` and `--tiller-namespace`
- logs for `draft logs` are stored in `$DRAFT_HOME/logs`
- moved `pkg/storage/kube` to `pkg/storage/kube/configmap` (this would be easy to revert/pull into a separate PR, but figured I should do it while in there)
- configuration for talking to docker has been moved to `draft up`

enables use cases like #564 

closes #548 
closes #546 
closes #517 
closes #508 
closes #485 
closes #482 
closes #479 
closes #470 
closes #428 
closes #404 
closes #356
closes #355 
closes #343 
closes #306 
closes #285 
closes #247 
closes #213 
closes #178 
closes #168 
closes #13

bacongobbler commented:
TODO:

- [ ] fix `draft logs
- [ ] update the reference docs
- [ ] find and remove dead code

bacongobbler commented:
Removing the WIP :)

squillace commented:
On Ubuntu 16.04 LTS, and I suspect any env on which docker requires `sudo` by default -- obviously NOT MAC WITH DOCKER FOR MAC :-) -- we either need to notice that, and request permission on init, or document that you need to remove sudo from docker using the instructions at https://docs.docker.com/install/linux/linux-postinstall/#manage-docker-as-a-non-root-user.

squillace commented:
Not a blocker, as this can be documented first, and the proper fix settled upon.

squillace commented:
Builds fine; pushes to docker hub fine (azurecr never finishes pushing; doesn't do minikube by default, per above), and I get this on the example-go app in the PR branch. It appears to be looking for the ingress in the values field.

```
╰─± d up
Draft Up Started: 'example-go'
example-go: Building Docker Image: SUCCESS ⚓  (1.0003s)
example-go: Pushing Docker Image: SUCCESS ⚓  (7.0016s)
buildApp: release error: could not install release: rpc error: code = Unknown desc = render error in "go/templates/ingress.yaml": template: go/templates/ingress.yaml:1:14: executing "go/templates/ingress.yaml" at <.Values.ingress.enab...>: can't evaluate field enabled in type interface {}
example-go: Releasing Application: FAIL ❌  (0.0280s)
example-go: Build ID: 01C8SB78P6GAXE4C2RQM4KJGVW
Inspect the logs with `draft logs 01C8SB78P6GAXE4C2RQM4KJGVW`
```

squillace commented:
Hmmm, a different error on this one. I create a new version of the same app, and I got this. Pushing to ACR using `az acr login` worked, whereas using `docker login` to ACR did not. but: 

```
╰─± d up
Draft Up Started: 'squillace/example-go-fooishness'
squillace/example-go-fooishness: Building Docker Image: SUCCESS ⚓  (4.0006s)
squillace/example-go-fooishness: Pushing Docker Image: SUCCESS ⚓  (16.0030s)
buildApp: release error: could not upgrade release: rpc error: code = Unknown desc = invalid release name, must match regex ^(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])+$ and the length must not longer than 53
squillace/example-go-fooishness: Releasing Application: FAIL ❌  (0.0194s)
squillace/example-go-fooishness: Build ID: 01C8SCCPXWH59VFBQ3NHKEG4EM
Inspect the logs with `draft logs 01C8SCCPXWH59VFBQ3NHKEG4EM`
```

bacongobbler commented:
> Not a blocker, as this can be documented first, and the proper fix settled upon.

Good feedback. I'll document this.

bacongobbler commented:
> Not clear how to map the local image repository to the cluster's repository. Does one, for example, need to use something like https://hub.docker.com/r/faithlife/minikube-registry-proxy/? or do the mapping manually, like in https://abhishek-tiwari.com/local-development-environment-for-kubernetes-using-minikube/?

It needs to be one registry that is accessible both to the user (draft) and the cluster (minikube). This is the same model prior to this PR, however this was made slightly easier because the docker daemon was co-located with the registry in Minikube's use case. We may need to revisit the local `registry` plugin for minikube and how we ineract with it, but I don't think it's something we need to solve right out of the gate.

> It appears to be looking for the ingress in the values field.

I made changes to the default packs that removes the ingress rules, but `draft init` will install packs from master, not this branch. When testing, you'll need to override the packs using the following commands:

```
draft pack-repo remove github.com/Azure/draft
draft pack-repo add https://github.com/bacongobbler/draft --version remove-draftd
```

As for this error:

> this seems to be because I named the image `squillace/example-go-fooishness`, and it blew on deploying something with a / in it.

I think you made the `name` field in draft.toml `squillace/example-go-fooishness`, right? If so you need to make that `example-go-fooishness` and the `repository` field set to `squillace.azurecr.io`. Does that help make more sense?

bacongobbler commented:
I just added `draft config get|set|unset|list`, so this now also closes #272 to address initial comments about being unable to set a global OS-wide registry for all apps deployed by a single tenant. I also dropped the `--registry` flag from `draft create` to pronounce that setting a local `registry` field in draft.toml is intended to override what's in `$DRAFT_HOME/config.toml`.

I'll see about getting this sucker rebased.

bacongobbler commented:
Ooh, forgot to add back the ingress rules. one sec

radu-matei commented:
One thing I would like to have is streaming logs (at least for the Docker build)

bacongobbler commented:
Would it be all right if we opened a new ticket for that? I'm just trying to bring this PR down to the bare essentials so any future enhancements can be built on top of this :)

bacongobbler commented:
action item:

- eliminate push step when on minikube a la #432

bacongobbler commented:
oh, I also filed #585 as a proposal to store build information locally. For now we'll switch over to the configmap backend so commands like `draft history` will continue to work with this new architecture.

bacongobbler commented:
last action item: need to read in docker envvars so draft can use the docker daemon via `eval $(minikube docker-env)`.

bacongobbler commented:
All right, so for anyone who is interested in testing this PR:

## Testing draftd-less Draft

1. check out the code and compile `draft` from source and move it somewhere on your $PATH, as usual
1.  `rm -rf ~/.draft`
1. Run `draft init` to re-create ~/.draft
1. Now, remove the "default" packs and replace with this branch's packs: `draft pack-repo remove github.com/Azure/draft` and `draft pack-repo add $GOPATH/src/github.com/Azure/draft`

Now you're ready to set up the cluster of your choice.

### Minikube Setup

1. If you have the registry add-on enabled, feel free to disable it as it's no longer necessary for Draft to work. `minikube addons disable registry` should do it.
1. It's easiest to delete your existing minikube cluster and start from a clean slate: `minikube delete`
1. `minikube start`
1. Once the cluster is up, run `eval $(minikube docker-env)` to populate your environment with the appropriate environment variables so Draft can use Minikube's Docker daemon. NOTE: you should not need to install the `docker` daemon or the CLI on your host for this to work. Draft should just pick this up automatically.
1. Run `helm init` to start Tiller

### Azure Setup

Start by creating a kubernetes cluster and a container registry in a supported region such as 'eastus':

```
az group create -n remove-draftd -g eastus
az aks create -g remove-draftd -n myk8scluster
az acr create -g remove-draftd -n myacr --sku Standard
```

Because we need to build images locally and push those images into the container registry, we need to [install `docker`](https://docs.docker.com/install).

Once docker is installed, we need to make sure the current user can access and run `docker` commands. On Ubuntu, this can be done by adding the current user to the `docker` group:

```
usermod -aG docker $(whoami)
```

Log out, then log back in as yourself for it to take effect.

Now we can log into the container registry:

```
az acr login -n myacr -g remove-draftd
```

And finally, tell Draft to push images to this registry:

```
draft config set registry myacr.azurecr.io
```

## Deploy an App

Now with the cluster of our choice set up, we can finally deploy an application.

```
$ cd examples/example-go
$ draft create
--> Draft detected Go (100.000000%)
--> Ready to sail
$ draft up
```

If anything isn't working according to the steps provided here, please let me know.

radu-matei commented on pkg/builder/builder.go:
```diff
@@ -0,0 +1,540 @@
+			fmt.Printf("buildApp: buildImg error: %v\n", err)
+			return
+		}
+		if app.ctx.Env.Registry != "" {
```
If we are not on Minikube and the registry is not set, this will skip the pushing step and will try to release the chart without the image being pushed to a registry.

The Draft output is `SUCCESS`, but it actually silently fails as it cannot find the image.

squillace replied:
oops

radu-matei replied:
Sample output:

```
Draft Up Started: 'example-go'
example-go: Building Docker Image: SUCCESS ⚓  (5.0139s)
example-go: Releasing Application: SUCCESS ⚓  (0.9168s)
example-go: Build ID: 01C950XBFNB14Y7CKQ90BHRRN9
Inspect the logs with `draft logs 01C950XBFNB14Y7CKQ90BHRRN9`
```

bacongobbler replied:
This is silently intentional. We have no way of detecting whether or not we are actually running against a non-minikube cluster. I've made a note of it in both the getting started guide as well as the reference docs, though I don't really see a way around this.

Any suggestions?

bacongobbler replied:
The hack I implemented long ago in `draft init` mightt work here (relying on the kubernetes config's current-context name to be "minikube", but it was more or less a crutch and probably won't hold up forever.

bacongobbler replied:
Would it be fair to inform the user that no registry was set before starting the build?

radu-matei replied:
The least we can do I think is outputting that the push step is skipped because there is no registry set in the config.

It would at least inform the user why.

Another thing that I have in mind - can we detect if it's Minikube by the master URL when using the Kubernetes client? It's surely localhost when Minikube, right?

squillace replied:
I don't like the "minikube" magic string solution, as I now have five different minikube profiles for different configurations, as you can imagine. I prefer to a) inform the person that no registry was set, so if it's not running against minikube it will skip the push activity and report success and yet fail; and b) doc the heck out of it.

However, another option might be to have the default be to push, but have the --skip-push flag such that when you know you're local, do this to skip push? Dunno. Whatchya think.

radu-matei replied:
Oh, GitHub refresh...

bacongobbler replied:
@squillace I considered the --skip-push flag, but that means minikube users now need to run `draft up --skip-push` all the time. It's a little more cumbersome for an "optional" feature flag, so I wanted to make the default experience sane.

I've made a change in 75fd33f so it'll display a warning, as well as a way to disable that warning if you know what you're doing. Sounds good?

squillace replied:
Let's try it for a bit, and if it's an issue we can identify a better ux. at this point, it's really a "did you remember to specify a registry if you're working against the cloud?" problem. It's not really a draft problem.

squillace commented:
On Ubuntu, rebuild for last commit, delete .draft, draft init, minikube: 

```
╰─± ../../bin/draft up
WARNING: no registry has been set, therefore Draft will not push to a container registry.
This can be fixed by running

        $ draft config set registry docker.io/myusername
Hint: this warning can be disabled by running

        $ draft config set disable-push-warning 1
Draft Up Started: 'example-go'
example-go: Building Docker Image -buildApp: buildImg error: Post http://192.168.99.100:2376/v1.23/build?buildargs=null&cachefrom=null&cgroupparent=&cpuperiod=0&cpuquota=0&cpusetcpus=&cpusetmems=&cpushares=0&dockerfile=&labels=null&memory=0&memswap=0&networkmode=&rm=0&shmsize=0&t=example-go%3A3a8762f94ec2e318dd55af9b08c9e53d345bd7c0&target=&ulimits=null: net/http: HTTP/1.x transport connection broken: malformed HTTP response "\x15\x03\x01\x00\x02\x02\x16".
* Are you trying to connect to a TLS-enabled daemon without TLS?
example-go: Building Docker Image: FAIL ❌  (1.0002s)
example-go: Build ID: 01C95ZESW75PVZEEMQ1CCBMPG3
Inspect the logs with `draft logs 01C95ZESW75PVZEEMQ1CCBMPG3`
```

I'll restart minikube.

squillace commented:
stopping, minikube delete, minikube start, rm .draft, eval step, draft init, helm init, same error.

commit:
commit e1a391394a426c0c6ebacafe8495605f2b7707fe
Author: Matthew Fisher <matt.fisher@microsoft.com>
Date:   Wed Mar 21 13:11:42 2018 -0700

    s/step/warning

bacongobbler commented:
Hmm, I hit that same error. I thought I fixed that earlier. Lemme try and debug the situation. Thanks for bringing that up!

bacongobbler commented:
ok, I think I fixed that up now. Feel free to give 'er another shot! :)

bacongobbler commented:
that was a doozie. Thanks to everyone involved for the feedback and reviews!

squillace commented:
ACK, @bacongobbler. If yoga boy got past that issue, then you don't have to wait for me, but I'll rebuild and retest.

bacongobbler commented:
Yoga boy did indeed get past this, haha :)

> you beat me to the punch with your last fix to the PR. But with that in, I have to express: https://gph.is/1r1DGY7 :boom:

squillace commented:
ETA for merge? want canary image built instantly as possible to start demo work with binaries. without devving fulltime, I lose track of branches/builds I did locally. :(bad PM, bad PM).

squillace commented:
me love me some demonless.
